be2-scala/src/test/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandlerSuite.scala

```diff
@@ -1,99 +1,119 @@
+import akka.actor.{ActorRef, ActorSystem, Props}
+import akka.pattern.AskableActorRef
 import akka.stream.scaladsl.{Flow, Sink, Source}
-import akka.testkit.{TestKit, TestProbe}
-import ch.epfl.pop.model.network.{JsonRpcResponse, ResultObject}
+import akka.testkit.TestKit
+import akka.util.Timeout
+import ch.epfl.pop.model.network.JsonRpcResponse
 import ch.epfl.pop.model.network.method.message.Message
-import ch.epfl.pop.model.objects.{Base64Data, Channel, DbActorNAckException, Hash}
-import ch.epfl.pop.pubsub.AskPatternConstants
-import ch.epfl.pop.pubsub.graph.{ErrorCodes, GraphMessage}
-import ch.epfl.pop.pubsub.graph.validators.RpcValidator
-import ch.epfl.pop.storage.DbActor
+import ch.epfl.pop.model.objects.Channel
+import ch.epfl.pop.pubsub.graph.GraphMessage
+import ch.epfl.pop.pubsub.{AskPatternConstants, MessageRegistry, PubSubMediator, PublishSubscribe}
+import ch.epfl.pop.storage.{DbActor, InMemoryStorage}
+import org.scalatest.BeforeAndAfterAll
 import org.scalatest.funsuite.AnyFunSuiteLike
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.matchers.should.Matchers.{convertToAnyShouldWrapper, equal}
 
 import scala.concurrent.Await
+import scala.concurrent.duration.{DurationInt, FiniteDuration}
+import scala.io.BufferedSource
+import scala.io.Source.fromFile
+import scala.util.Success
 
-class GetMessagesByIdResponseHandlerSuite extends TestKit(ActorSystem("GetMessagesByIdResponseHandlerSuiteSystem")) with AnyFunSuiteLike with AskPatternConstants {
+class GetMessagesByIdResponseHandlerSuite extends TestKit(ActorSystem("GetMessagesByIdResponseHandlerSuiteSystem")) with AnyFunSuiteLike with AskPatternConstants with BeforeAndAfterAll {
 
-  class TestDb(testProbe: ActorRef) extends Actor {
-    override def receive: Receive = {
-      case DbActor.WriteAndPropagate(channel, message) =>
-        testProbe ! (channel, message)
-        sender() ! DbActor.DbActorAck()
+  // Implicit for system actors
+  implicit val timeout: Timeout = Timeout(1.seconds)
+
+  val MAX_TIME: FiniteDuration = 2 * duration
+  val inMemoryStorage: InMemoryStorage = InMemoryStorage()
+  val messageRegistry: MessageRegistry = MessageRegistry()
+  val pubSubMediatorRef: ActorRef = system.actorOf(PubSubMediator.props, "PubSubMediator")
+  val dbActorRef: AskableActorRef = system.actorOf(Props(DbActor(pubSubMediatorRef, messageRegistry, inMemoryStorage)), "DbActor")
+
+  // Inject dbActor above
+  PublishSubscribe.buildGraph(pubSubMediatorRef, dbActorRef, messageRegistry, ActorRef.noSender, ActorRef.noSender, isServer = false)
+
+  // handler we want to test
+  val responseHandler: Flow[GraphMessage, GraphMessage, NotUsed] = GetMessagesByIdResponseHandler.responseHandler(MessageRegistry())
+
+  // loading the valid json
+  val sourceCorrectGetMessageById: BufferedSource =
+    fromFile("src/test/scala/util/examples/json/get_messages_by_id_answer.json")
+
+  val linesCorrectJson: String =
```

Comment:
Why do you need this step ? You cannot directly use the `sourceCorrectGetMessageById` ?

Reply:
We have to read it no matter what, and if we read it directly by calling` .mkString` then we lose the source reference and cannot close the file. I added a private helper function to avoid code repetition though. If we realise we read lots of jsons file it might be worth to move that function somewhere accessible but for now its not really worth it. 

cf commit: 151953a097d25a083c2011a79ef54b4aa69a8660